File: frontend/src/lib/languages.ts
export type AppLanguage = {
  name: string;
  code: string;
  native: string;
  dir: "ltr" | "rtl";
};

/** Keep in sync with backend/app/i18n/languages.py */
export const GROQ_SMALL_MULTILINGUAL_MODEL = "openai/gpt-oss-20b";

export const SUPPORTED_LANGUAGES: AppLanguage[] = [
  { name: "English", code: "en", native: "English", dir: "ltr" },
  { name: "Hindi", code: "hi", native: "हिन्दी", dir: "ltr" },
  { name: "Hinglish", code: "hi-Latn", native: "Hinglish", dir: "ltr" },
  { name: "Telugu", code: "te", native: "తెలుగు", dir: "ltr" },
  { name: "Urdu", code: "ur", native: "اردو", dir: "rtl" },
];

export const DEFAULT_LANGUAGE = "English";

const STORAGE_KEY = "resq-app-language";

export function languageByName(name: string): AppLanguage {
  return (
    SUPPORTED_LANGUAGES.find((item) => item.name.toLowerCase() === name.toLowerCase()) ??
    SUPPORTED_LANGUAGES[0]
  );
}

export function normalizeLanguage(value: string | null | undefined): string {
  if (!value?.trim()) return DEFAULT_LANGUAGE;
  const raw = value.trim().toLowerCase();
  const aliases: Record<string, string> = {
    en: "English",
    hi: "Hindi",
    hin: "Hindi",
    "hi-latn": "Hinglish",
    hinglish: "Hinglish",
    "hi-en": "Hinglish",
    te: "Telugu",
    ur: "Urdu",
  };
  if (aliases[raw]) return aliases[raw];
  const match = SUPPORTED_LANGUAGES.find((item) => item.name.toLowerCase() === raw);
  return match?.name ?? DEFAULT_LANGUAGE;
}

export function readStoredLanguage(): string {
  if (typeof window === "undefined") return DEFAULT_LANGUAGE;
  return normalizeLanguage(window.localStorage.getItem(STORAGE_KEY));
}

export function storeLanguage(name: string) {
  const normalized = normalizeLanguage(name);
  if (typeof window !== "undefined") {
    window.localStorage.setItem(STORAGE_KEY, normalized);
  }
  return normalized;
}

export function languageGroups() {
  return { india: SUPPORTED_LANGUAGES, international: [] as AppLanguage[] };
}

File: backend/app/i18n/languages.py
"""Languages used with the Groq chat model and the ResQ UI.

For now the product exposes English, Hindi, Urdu, Telugu, and Hinglish.
Groq still receives the human-readable name (plus a Hinglish style note).
Whisper uses ISO-639-1; Hinglish maps to Hindi.
"""

from __future__ import annotations

from typing import Any

GROQ_SMALL_MULTILINGUAL_MODEL = "openai/gpt-oss-20b"

# Whisper ISO-639-1 codes. Name is what the client sends as `language`.
SUPPORTED_LANGUAGES: tuple[dict[str, str], ...] = (
    {"name": "English", "code": "en", "native": "English", "group": "india"},
    {"name": "Hindi", "code": "hi", "native": "हिन्दी", "group": "india"},
    {"name": "Hinglish", "code": "hi-Latn", "native": "Hinglish", "group": "india"},
    {"name": "Telugu", "code": "te", "native": "తెలుగు", "group": "india"},
    {"name": "Urdu", "code": "ur", "native": "اردو", "group": "india"},
)

DEFAULT_LANGUAGE = "English"

_ALIASES = {
    "en-in": "en",
    "eng": "en",
    "hin": "hi",
    "hi-en": "hi-latn",
    "hinglish": "hi-latn",
    "hi-latn": "hi-latn",
    "telugu": "te",
    "urd": "ur",
}

_BY_NAME = {item["name"].lower(): item for item in SUPPORTED_LANGUAGES}
_BY_CODE = {item["code"].lower(): item for item in SUPPORTED_LANGUAGES}


def language_catalog() -> dict[str, Any]:
    return {
        "groq_model": GROQ_SMALL_MULTILINGUAL_MODEL,
        "default_language": DEFAULT_LANGUAGE,
        "languages": [dict(item) for item in SUPPORTED_LANGUAGES],
    }


def _lookup(raw: str) -> dict[str, str] | None:
    key = raw.strip().lower()
    key = _ALIASES.get(key, key)
    return _BY_NAME.get(key) or _BY_CODE.get(key)


def normalize_language(value: str | None) -> str:
    if not value or not str(value).strip():
        return DEFAULT_LANGUAGE
    item = _lookup(str(value))
    return item["name"] if item else DEFAULT_LANGUAGE


def whisper_code(value: str | None) -> str | None:
    if not value:
        return None
    item = _lookup(str(value))
    if not item:
        return None
    code = item["code"].lower()
    if code.startswith("hi"):
        return "hi"
    return item["code"]


def groq_language_phrase(value: str | None) -> str:
    name = normalize_language(value)
    if name == "Hinglish":
        return (
            "Hinglish ONLY — Hindi+English mix in Latin letters "
            "(example: 'Aadmi gadi se gir gaya aur khoon nikal raha hai. Weight mat dalo.'). "
            "Do not write full English sentences. Do not use Devanagari or Urdu script."
        )
    if name == "Hindi":
        return "Hindi only, in Devanagari script. Do not reply in English."
    if name == "Telugu":
        return "Telugu only, in Telugu script. Do not reply in English."
    if name == "Urdu":
        return "Urdu only, in Urdu/Arabic script. Do not reply in English."
    return "English"


def is_english_reply(value: str | None) -> bool:
    return normalize_language(value) == "English"


def language_map() -> dict[str, str]:
    mapping: dict[str, str] = {}
    for item in SUPPORTED_LANGUAGES:
        mapping[item["name"].lower()] = whisper_code(item["name"]) or item["code"]
        mapping[item["code"].lower()] = whisper_code(item["name"]) or item["code"]
    mapping["hinglish"] = "hi"
    mapping["hi-latn"] = "hi"
    return mapping
